<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Running zAMP &#8212; zAMP v.1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=6c7b0eaa"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Under the hood" href="under_the_hood.html" />
    <link rel="prev" title="Taxonomic reference databases" href="ref_DB_preprocessing.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="running-zamp">
<span id="execution"></span><h1>Running zAMP<a class="headerlink" href="#running-zamp" title="Link to this heading">¶</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before running zAMP see <a class="reference internal" href="setup.html#setup"><span class="std std-ref">Installation and resource requirements</span></a> and <a class="reference internal" href="ref_DB_preprocessing.html#tax-db"><span class="std std-ref">Taxonomic reference databases</span></a> sections.</p>
</div>
<p>zAMP can use reads locally or from NCBI’s <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive (SRA)</a>.</p>
<section id="local-reads">
<h2>Local reads<a class="headerlink" href="#local-reads" title="Link to this heading">¶</a></h2>
<section id="sample-sheet-only">
<h3>Sample sheet only<a class="headerlink" href="#sample-sheet-only" title="Link to this heading">¶</a></h3>
<p><em>Command</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">run</span> <span class="o">-</span><span class="n">i</span> <span class="n">samples</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p><em>Example samples.tsv</em> :</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>R1</p></th>
<th class="head"><p>R2</p></th>
<th class="head"><p>sample_group</p></th>
<th class="head"><p>run</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SRR9067116</p></td>
<td><p>reads/SRR9067116_1.fastq.gz</p></td>
<td><p>reads/SRR9067116_2.fastq.gz</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-odd"><td><p>SRR9067115</p></td>
<td><p>reads/SRR9067115_1.fastq.gz</p></td>
<td><p>reads/SRR9067115_2.fastq.gz</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-even"><td><p>SRR9067114</p></td>
<td><p>reads/SRR9067114_1.fastq.gz</p></td>
<td><p>reads/SRR9067114_2.fastq.gz</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-odd"><td><p>SRR7225909</p></td>
<td><p>reads/SRR7225909_1.fastq.gz</p></td>
<td><p>reads/SRR7225909_2.fastq.gz</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
<tr class="row-even"><td><p>SRR7225908</p></td>
<td><p>reads/SRR7225908_1.fastq.gz</p></td>
<td><p>reads/SRR7225908_2.fastq.gz</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
<tr class="row-odd"><td><p>SRR7225907</p></td>
<td><p>reads/SRR7225907_1.fastq.gz</p></td>
<td><p>reads/SRR7225907_2.fastq.gz</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><cite>sample</cite>: the name of the sample</p></li>
<li><p><cite>R1</cite>: path to forward reads</p></li>
<li><p><cite>R2</cite>: path to reverse reads</p></li>
<li><p><cite>sample_group</cite>: sample grouping for visualizations</p></li>
<li><p><cite>run</cite>: column for applying DADA2 error learning and denoising for each sequencing run</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can add any other columns in the table provided the above mentionned columns are present</p>
</div>
</section>
<section id="reads-directory-and-metadata-as-input">
<h3>Reads directory and metadata as input<a class="headerlink" href="#reads-directory-and-metadata-as-input" title="Link to this heading">¶</a></h3>
<p><em>Command</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">run</span> <span class="o">-</span><span class="n">i</span> <span class="n">reads</span> <span class="o">-</span><span class="n">m</span> <span class="n">metadata</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p><em>Example metadata.tsv</em> :</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>fastq</p></th>
<th class="head"><p>sample</p></th>
<th class="head"><p>sample_group</p></th>
<th class="head"><p>run</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SRR9067116</p></td>
<td><p>Vaginal-Library42</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-odd"><td><p>SRR9067115</p></td>
<td><p>Vaginal-Library41</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-even"><td><p>SRR9067114</p></td>
<td><p>Vaginal-Library48</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
</tr>
<tr class="row-odd"><td><p>SRR7225909</p></td>
<td><p>NE14</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
<tr class="row-even"><td><p>SRR7225908</p></td>
<td><p>A3D12</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
<tr class="row-odd"><td><p>SRR7225907</p></td>
<td><p>NN15</p></td>
<td><p>human_biliary_tract</p></td>
<td><p>run2</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="sra-reads">
<h2>SRA reads<a class="headerlink" href="#sra-reads" title="Link to this heading">¶</a></h2>
<p>zAMP can fetch reads from NCBI’s <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra">SRA</a> using <a class="reference external" href="https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump">fasterq-dump</a>.</p>
<p><em>Command</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">run</span> <span class="o">-</span><span class="n">i</span> <span class="n">sra</span><span class="o">-</span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p><em>Example sra-samples.tsv</em> :</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>sample_label</p></th>
<th class="head"><p>sample_group</p></th>
<th class="head"><p>run</p></th>
<th class="head"><p>paired</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>SRR9067116</p></td>
<td><p>Vaginal-16s-V3V4-Library42</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>SRR9067115</p></td>
<td><p>Vaginal-16s-V3V4-Library41</p></td>
<td><p>Genital_tract</p></td>
<td><p>run1</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><cite>sample_label</cite>: label to rename the SRA fastq files</p></li>
<li><p><cite>paired</cite>: whether reads are paired or not (required because snakemake can’t guess the pairing from the fastq outputs easily )</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">zAMP</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Installation and resource requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="ref_DB_preprocessing.html">Taxonomic reference databases</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running zAMP</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#local-reads">Local reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sra-reads">SRA reads</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="under_the_hood.html">Under the hood</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream_analysis.html">Downstream Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="insilico_validation.html"><em>In silico</em> validation tool</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="ref_DB_preprocessing.html" title="previous chapter">Taxonomic reference databases</a></li>
      <li>Next: <a href="under_the_hood.html" title="next chapter">Under the hood</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2020, MetaGenLab.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/pages/execution.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>