<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>In silico validation tool &#8212; zAMP v.1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=6c7b0eaa"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Frequently asked questions (FAQ)" href="FAQ.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="in-silico-validation-tool">
<span id="insilico"></span><h1><em>In silico</em> validation tool<a class="headerlink" href="#in-silico-validation-tool" title="Link to this heading">¶</a></h1>
<section id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Link to this heading">¶</a></h2>
<p>This module aims at predicting <em>in silico</em> if specific taxa are:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>amplified by a set of PCR primers used for amplicon-based metagenomics</p></li>
<li><p>accurately classified taxonomically based on the generated amplicon</p></li>
</ol>
</div></blockquote>
</section>
<section id="working-principle">
<h2>Working principle<a class="headerlink" href="#working-principle" title="Link to this heading">¶</a></h2>
<p>Based on a user-defined list of NCBI tax IDs, assemblies or taxon queries, genome assemblies are downloaded from the NCBI database with <a class="reference external" href="https://github.com/metagenlab/assembly_finder">Assembly Finder</a>. Then, PCR primer sequences provided by the user are used to run an <em>in silico</em> PCR with <a class="reference external" href="https://github.com/egonozer/in_silico_pcr">in_silico_pcr</a> (or alternatively, with <a class="reference external" href="https://github.com/metagenlab/updated_simulate_PCR">simulate_PCR</a>). The generated <em>in silico</em> amplicons are  treated by the main pipeline as they would if they were the results of sequencing reads (primer trimming, amplicon clustering into representative sequences, taxonomic classification).</p>
<p>Finally, for each of the downloaded assembly, this module provides a table with a description of the amplicons predicted to be amplified with the PCR primers (number of sequence variants, number of copies) as well as the expected and obtained taxonomic assignment.</p>
</section>
<section id="inputs">
<h2>Inputs<a class="headerlink" href="#inputs" title="Link to this heading">¶</a></h2>
<p>To execute the pipeline, one needs:</p>
<ul class="simple">
<li><p>An input file containing the accession names or the Tax IDs of interest. This is a one-column text file without headers. The identifiers should match NCBI taxonomy. One can skip this text file and use a query term instead, see usage cases below.</p></li>
<li><p><span class="xref std std-ref">A taxonomic database preprocessed with our dedicated pipeline</span></p></li>
</ul>
<p><strong>Input file example:</strong></p>
<ul class="simple">
<li><p>With accession names:</p></li>
</ul>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>GCA_000008005.1
GCA_000010425.1
GCA_000016965.1
GCA_020546685.1
GCA_000172575.2
GCA_000005845.2
GCA_000014425.1
GCA_003324715.1
GCA_000007645.1
GCA_000007465.2
GCA_013372085.1
GCA_031191545.1
GCA_000012825.1
GCA_000307795.1
GCA_000008805.1
GCA_000010505.1
GCA_000231215.1
GCA_000017205.1
GCA_000013425.1
GCA_000007265.1
</pre></div>
</div>
<ul class="simple">
<li><p>With NCBI tax IDs:</p></li>
</ul>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>1069201
182096
41058
1220207
1287682
746128
5059
5062
</pre></div>
</div>
</section>
<section id="execution">
<h2>Execution<a class="headerlink" href="#execution" title="Link to this heading">¶</a></h2>
<p>The module is executed with <cite>zamp insilico</cite>.
You can see all required and optional arguments with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">insilico</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>Example usage cases:</p>
<ul>
<li><p>Using bacteria assembly accession names (note the –accession argument when using accession names instead of tax IDs):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">insilico</span> <span class="o">-</span><span class="n">i</span> <span class="n">zamp</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">bacteria</span><span class="o">-</span><span class="n">accs</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">db</span> <span class="n">greengenes2</span> <span class="o">--</span><span class="n">accession</span> <span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CCTACGGGNGGCWGCAG</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">GACTACHVGGGTATCTAATCC</span>
</pre></div>
</div>
</li>
<li><p>Using fungi tax IDs (requires additional ITS amplicon-specific parameters to adjust the amplicon size)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">insilico</span> <span class="o">-</span><span class="n">i</span> <span class="n">zamp</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">fungi</span><span class="o">-</span><span class="n">taxa</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">-</span><span class="n">db</span> <span class="n">unite_db_v10</span> \
<span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CYHRGYYATTTAGAGGWMSTAA</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">RCKDYSTTCWTCRWYGHTGB</span> \
<span class="o">--</span><span class="n">minlen</span> <span class="mi">50</span> <span class="o">--</span><span class="n">maxlen</span> <span class="mi">900</span>
</pre></div>
</div>
</li>
<li><p>Using a query term. In this example, 100 assemblies will be downloaded per taxon (<code class="docutils literal notranslate"><span class="pre">nb</span> <span class="pre">100</span></code>) including non-reference assemblies (<cite>not-only-ref</cite>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">insilico</span> <span class="o">-</span><span class="n">i</span> <span class="s2">&quot;lactobacillus&quot;</span> \
<span class="o">-</span><span class="n">db</span> <span class="n">ezbiocloud</span> \
<span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CCTACGGGNGGCWGCAG</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">GACTACHVGGGTATCTAATCC</span> \
<span class="o">--</span><span class="n">replace</span><span class="o">-</span><span class="n">empty</span> <span class="o">-</span><span class="n">nb</span> <span class="mi">100</span> <span class="o">--</span><span class="ow">not</span><span class="o">-</span><span class="n">only</span><span class="o">-</span><span class="n">ref</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">¶</a></h2>
<p>The pipeline gathers information on available assemblies for the requested taxIDs in the <cite>assembly_finder</cite> folder.</p>
<p>The output of the in-silico amplification is in <cite>Insilico</cite> folder, and contains the following subfolders:</p>
<ul class="simple">
<li><p>PCR: contains the output of in-silico PCR amplification</p></li>
<li><p>2_denoised: output of clustering and denoising into representative sequences, and count tables</p></li>
<li><p>3_classified: output of taxonomic classification and tables comparing expected and obtained taxonomic assignations (<cite>InSilico_compare_tax.tsv</cite> and <cite>InSilico_compare_tax_long.tsv</cite>.)</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">zAMP</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Installation and resource requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="ref_DB_preprocessing.html">Taxonomic reference databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Running zAMP</a></li>
<li class="toctree-l1"><a class="reference internal" href="under_the_hood.html">Under the hood</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream_analysis.html">Downstream Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#"><em>In silico</em> validation tool</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#aim">Aim</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-principle">Working principle</a></li>
<li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#execution">Execution</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="FAQ.html" title="previous chapter">Frequently asked questions (FAQ)</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2020, MetaGenLab.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/pages/insilico_validation.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>