<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Taxonomic reference databases &#8212; zAMP v.1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=6c7b0eaa"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script type="module" src="https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs"></script>
    <script type="module" src="https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs"></script>
    <script type="module">import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";import elkLayouts from "https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs";mermaid.registerLayoutLoaders(elkLayouts);mermaid.initialize({startOnLoad:false});</script>
    <script src="https://cdn.jsdelivr.net/npm/d3@7.9.0/dist/d3.min.js"></script>
    <script type="module">
import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";
window.addEventListener("load", () => mermaid.run());
</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pipeline execution" href="execution.html" />
    <link rel="prev" title="Installation and resource requirements" href="setup.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="taxonomic-reference-databases">
<span id="tax-db"></span><h1>Taxonomic reference databases<a class="headerlink" href="#taxonomic-reference-databases" title="Link to this heading">¶</a></h1>
<section id="database-processing-principle">
<h2>Database processing principle<a class="headerlink" href="#database-processing-principle" title="Link to this heading">¶</a></h2>
<p>zAMP offers the possibility to train classifiers like <em>RDP</em> <a class="footnote-reference brackets" href="#id14" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>, <em>QIIME</em> <a class="footnote-reference brackets" href="#id15" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a> and <em>Decipher</em> <a class="footnote-reference brackets" href="#id16" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a> on full length sequences or subdomains like the V3-V4 region in the 16S rRNA gene.</p>
<p>Classification accuracy depends on classifiers parameters and the region on which they were trained on <a class="footnote-reference brackets" href="#id17" id="id4" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>. In fact, Bokulich et al. <a class="footnote-reference brackets" href="#id17" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a> demonstrated that training classifiers on specific regions leads to enhanced accuracy compared to using the full length sequences (for short reads).</p>
<p>Therefore, by default, zAMP extracts primer amplified regions with cutadapt <a class="footnote-reference brackets" href="#id18" id="id6" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>, dereplicates and clusters sequences with vseach <a class="footnote-reference brackets" href="#id19" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a>, and adapts taxonomy according to these clusters.</p>
<p>Database processing flowchart:</p>
<pre align="center" class="mermaid align-center">
        flowchart TD
A[&quot;dna-sequences.fasta&quot;]
taxonomy[&quot;taxonomy.tsv&quot;]
taxonomy --&gt; clean(&quot;Clean taxonomy&quot;)
clean --&gt; B
A --&gt; B{Extract region ?}
B --&gt; |Yes| C(&quot;Extract regions
(cutadapt)&quot;)
B --&gt; |No| D(Copy files)
C --&gt; E(&quot;Cluster
(vsearch)&quot;)
E --&gt; F(&quot;Derep and merge taxonomy&quot;)
F --&gt; G(&quot;Train classifiers
(RDP, Decipher...) &quot;)
    </pre></section>
<section id="custom-database">
<h2>Custom database<a class="headerlink" href="#custom-database" title="Link to this heading">¶</a></h2>
<p>Taxonomy table in QIIME format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">1</span>   <span class="n">Bacteria</span><span class="p">;</span><span class="n">Proteobacteria</span><span class="p">;</span><span class="n">Alphaproteobacteria</span><span class="p">;</span><span class="n">Rhodospirillales</span><span class="p">;</span><span class="n">Rhodospirillaceae</span><span class="p">;</span><span class="n">Magnetospirillum</span><span class="p">;</span><span class="n">Magnetospirillum</span> <span class="n">magnetotacticum</span>
<span class="mi">2</span>   <span class="n">Bacteria</span><span class="p">;</span><span class="n">Fusobacteria</span><span class="p">;</span><span class="n">Fusobacteria_c</span><span class="p">;</span><span class="n">Fusobacteriales</span><span class="p">;</span><span class="n">Fusobacteriaceae</span><span class="p">;</span><span class="n">Fusobacterium</span><span class="p">;</span><span class="n">Fusobacterium</span> <span class="n">nucleatum</span>
</pre></div>
</div>
<p>Fasta file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="mi">1</span>
<span class="n">CTGNCGGCGTGCCTAACACATNCAAGTCGAGCGGTGCTACGGAGGTCTTCGGACTGAAGTAGCATAGCGGCGGACGGGTGAGTAATACACAGGAACGTGCCCCTTGGAGGCGGATAGCTGTGGGAAACTGCAGGTAATCCGCCGTAAGCTCGGGAGAGGAAAGCCGGAAGGCGCCGAGGGAGCGGCCTGTGGCCCATCAGGTAGTTGGTAGGGTAAGAGCCTACCAAGCCGACGACGGGTAGCCGGTCTGAGAGGATGGACGGCCACAAGGGCACTGAGACACGGGCCCTACTCCTACGGGAGGCAGCAGTGGGGGATATTGGACAATGGGCGAAAGCCTGATCCAGCGACGCCGCGTGAGGGACGAAGTCCTTCGGGACGTAAACCTCTGTTGTAGGGGAAGAAGACAGTGACGGTACCCTACGAGGAAGCCCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGGNCGAGCGTTACCCGGAATCACTGGGCGTAAAGGGTGCGTA</span>
<span class="o">&gt;</span><span class="mi">2</span>
<span class="n">AACGAACGCTGGCGGCAGGCTTAACACATGCAAGTCGAACGAAGTCTTCGGACTTAGTGGCGCACGGGTGAGTAACACGTGGGAATATACCTCTTGGTGGGGAATAACGTCGGGAAACTGACGCTAATACCGCATACGCCCTTCGGGGGAAAGATTTATCGCCGAGAGATTAGCCCGCGTCCGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCGGTAGCTGGTCTGAGAGGATGATCAGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGAAAGCCTGATCCAGCCATGCCGCGTGAGTGATGAAGGCCTTAGGGTTGTAAAGCTCTTTCACCCACGACGATGATGACGGTAGTGGGAGAAGAAGCCCCGGCTAACTTCGTGCCAGCAGCCGCGGTAATACGAAGGGGGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTGGTCATAGTCAGAAGTGAAAGCCCTGGGCTCAACCCGGGAATTGCTTTTGATACTGGACCGCTAGAATCACGGAGAGGGTAGTGGAATTCCGAGTGTAGAGGTGAAATTCGTAGATATTCGGAAGAACACCAGTGGCGAAGG</span>
</pre></div>
</div>
<p>Default command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> <span class="o">--</span><span class="n">taxonomy</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">tsv</span> \
        <span class="o">--</span><span class="n">fasta</span> <span class="n">sequences</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CCTACGGGNGGCWGCAG</span> \
        <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">GACTACHVGGGTATCTAATCC</span> \
        <span class="o">-</span><span class="n">o</span> <span class="n">processed</span><span class="o">-</span><span class="n">db</span>
</pre></div>
</div>
<p>Skip primer amplified region extraction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> <span class="o">--</span><span class="n">taxonomy</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">tsv</span> \
    <span class="o">--</span><span class="n">fasta</span> <span class="n">sequences</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">processing</span> \
    <span class="o">-</span><span class="n">o</span> <span class="n">unprocessed</span><span class="o">-</span><span class="n">db</span>
</pre></div>
</div>
</section>
<section id="available-databases">
<h2>Available databases<a class="headerlink" href="#available-databases" title="Link to this heading">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Processed and unprocessed SILVA, Greengenes2 and UNITE will be made available soon</p>
</div>
<p>Here is a short, non-exhaustive, list of databases from which we could successfully prepare a database:</p>
<ul>
<li><p>EzBioCloud (16S rRNA  - Bacteria)</p>
<blockquote>
<div><p><a class="reference external" href="https://www.ezbiocloud.net/resources/16s_download">Website</a></p>
<p><a class="reference external" href="https://doi.org/10.1099/ijsem.0.001755">Publication</a></p>
</div></blockquote>
</li>
<li><p>SILVA (16/18S rRNA, 23/28S rRNA - Bacteria and Eukarya )</p>
<blockquote>
<div><p><a class="reference external" href="https://www.arb-silva.de/download/arb-files/">Website</a></p>
<p><a class="reference external" href="https://doi.org/10.1093/nar/gks1219">Publication</a></p>
</div></blockquote>
</li>
<li><p>UNITE (ITS - Eukarya)</p>
<blockquote>
<div><p><a class="reference external" href="https://unite.ut.ee/repository.php">Website</a>
<a class="reference external" href="https://doi.org/10.1093/nar/gkad1039">Publication</a></p>
</div></blockquote>
</li>
<li><p>Eukaryome (ITS - Eukarya)</p>
<blockquote>
<div><p><a class="reference external" href="https://eukaryome.org/download/">Website</a>
<a class="reference external" href="https://doi.org/10.1093/database/baae043">Publication</a></p>
</div></blockquote>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For Eukaryome, additional steps might be needed to filter a kingdom of interest (e.g. Fungi), or remove entries with incomplete taxonomy.</p>
</div>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Link to this heading">¶</a></h2>
<p>Command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<ul class="simple">
<li><p>The “–no-processing” parameter enables to skip the preprocessing and only format the provided database and train the classifiers.</p></li>
<li><p>“fw-primer” and “rv-primer” are fed to <a class="reference external" href="https://cutadapt.readthedocs.io/en/v3.0/guide.html#linked-adapters-combined-5-and-3-adapter">cutadapt linked adapter argument</a>.</p></li>
<li><p>“–cutadapt_args_fw” and “–cutadapt_args_rv” allow to pass additional arguments to cutadapt, affecting the forward and reverse primer, respectively. It for instance allows to indicate which primer is optional &lt;<a class="reference external" href="https://cutadapt.readthedocs.io/en/v3.0/guide.html#changing-which-adapters-are-required">https://cutadapt.readthedocs.io/en/v3.0/guide.html#changing-which-adapters-are-required</a>&gt;`_. It is particularly useful when trying to extract ITS1 amplicons: the 5’ universal primer is located on the SSU rRNA preceding the ITS region and thus is absent in ITS reference database. In this case, providing “–cutadapt_args_fw optional” enables to make it optional.</p></li>
<li><p>“errors” is fed to <a class="reference external" href="https://cutadapt.readthedocs.io/en/v3.0/guide.html#minimum-overlap-reducing-random-matches">cutadapt to define the number of accepted mismatches per primer</a>.</p></li>
<li><p>“ampcov” is used with the length of the provided primers to feed <a class="reference external" href="https://cutadapt.readthedocs.io/en/v3.0/guide.html#minimum-overlap-reducing-random-matches">cutadapt with a minimal overlap</a>.</p></li>
</ul>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading">¶</a></h2>
<section id="bacteria">
<h3>Bacteria<a class="headerlink" href="#bacteria" title="Link to this heading">¶</a></h3>
<p><strong>Greengenes2</strong></p>
<ul>
<li><p>Download:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">microbio</span><span class="o">.</span><span class="n">me</span><span class="o">/</span><span class="n">greengenes_release</span><span class="o">/</span><span class="mf">2022.10</span><span class="o">/</span><span class="mf">2022.10</span><span class="o">.</span><span class="n">backbone</span><span class="o">.</span><span class="n">full</span><span class="o">-</span><span class="n">length</span><span class="o">.</span><span class="n">fna</span><span class="o">.</span><span class="n">qza</span> <span class="o">&amp;&amp;</span> \
<span class="n">wget</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">microbio</span><span class="o">.</span><span class="n">me</span><span class="o">/</span><span class="n">greengenes_release</span><span class="o">/</span><span class="mf">2022.10</span><span class="o">/</span><span class="mf">2022.10</span><span class="o">.</span><span class="n">backbone</span><span class="o">.</span><span class="n">tax</span><span class="o">.</span><span class="n">qza</span>
</pre></div>
</div>
</li>
<li><p>Decompress qza with <a class="reference external" href="https://docs.qiime2.org/2024.5/tutorials/exporting/)">qiime2 export</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>docker run -t -i -v $(pwd):/data quay.io/qiime2/tiny:2024.5 \
qiime tools export \
--input-path 2022.10.backbone.full-length.fna.qza \
--output-path greengenes2 &amp;&amp; \
docker run -t -i -v $(pwd):/data quay.io/qiime2/tiny:2024.5 \
qiime tools export \
--input-path 2022.10.backbone.tax.qza --output-path greengenes2
</pre></div>
</div>
</li>
<li><p>Prepare database:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> <span class="o">--</span><span class="n">fasta</span> <span class="n">greengenes2</span><span class="o">/</span><span class="n">dna</span><span class="o">-</span><span class="n">sequences</span><span class="o">.</span><span class="n">fasta</span> \
<span class="o">--</span><span class="n">taxonomy</span> <span class="n">greengenes2</span><span class="o">/</span><span class="n">taxonomy</span><span class="o">.</span><span class="n">tsv</span> <span class="o">--</span><span class="n">name</span> <span class="n">greengenes2</span> \
<span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CCTACGGGNGGCWGCAG</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">GACTACHVGGGTATCTAATCC</span> \
<span class="o">-</span><span class="n">o</span> <span class="n">greengenes2</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="fungi">
<h3>Fungi<a class="headerlink" href="#fungi" title="Link to this heading">¶</a></h3>
<p><strong>Unite ITS1</strong></p>
<p>Fungal ITS databases Unite v10 and Eukaryome v1.8 do not contain the adjacent SSU/LSU sequences (they contain 5.8S), where some of the commonly used PCR primers lie on.
It is important to adjust the cutadapt parameters so that only the absent primer is optional.
In the following example, we prepare a database for fungal ITS1 from Unite Db.
In this case, the forward primer (lying of the 18S) will not be present in most sequences of Unite/Eukaryome (but the reverse primer lying on the 5.8S is present); therefore we set the forward primer as optional; the extracted sequences will start at the available 5’ of the database and end at the reverse primer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> \
<span class="o">--</span><span class="n">fasta</span> <span class="n">sh_refs_qiime_unite_ver10_dynamic_04</span><span class="mf">.04.2024</span><span class="o">.</span><span class="n">fasta</span> \
<span class="o">--</span><span class="n">taxonomy</span> <span class="n">sh_taxonomy_qiime_unite_ver10_dynamic_04</span><span class="mf">.04.2024</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">name</span> <span class="n">unite</span> \
<span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">CYHRGYYATTTAGAGGWMSTAA</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">RCKDYSTTCWTCRWYGHTGB</span> \
<span class="o">--</span><span class="n">minlen</span> <span class="mi">50</span> <span class="o">--</span><span class="n">maxlen</span> <span class="mi">900</span> \
<span class="o">--</span><span class="n">cutadapt_args_fw</span> <span class="s2">&quot;optional&quot;</span> \
<span class="o">-</span><span class="n">o</span> <span class="n">unite_ITS1</span>
</pre></div>
</div>
<p><strong>Eukaryome ITS2</strong></p>
<p>Similarly, to extract ITS2 from fungal databases such as Eukaryome, the reverse primer needs to be set as optional, because it is located on the LSU, which is absent in the database sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp</span> <span class="n">db</span> \
<span class="o">--</span><span class="n">fasta</span> <span class="n">QIIME2_EUK_ITS_v1</span><span class="mf">.8</span><span class="o">.</span><span class="n">fasta</span> \
<span class="o">--</span><span class="n">taxonomy</span> <span class="n">QIIME2_EUK_ITS_v1</span><span class="mf">.8</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">name</span> <span class="n">eukaryome</span> \
<span class="o">--</span><span class="n">fw</span><span class="o">-</span><span class="n">primer</span> <span class="n">GCATCGATGAAGAACGCAGC</span> <span class="o">--</span><span class="n">rv</span><span class="o">-</span><span class="n">primer</span> <span class="n">TCCTCCGCTTATTGATATGC</span> \
<span class="o">--</span><span class="n">minlen</span> <span class="mi">50</span> <span class="o">--</span><span class="n">maxlen</span> <span class="mi">900</span> \
<span class="o">--</span><span class="n">cutadapt_args_rv</span> <span class="s2">&quot;optional&quot;</span> \
<span class="o">-</span><span class="n">o</span> <span class="n">eukaryome_ITS2</span>
</pre></div>
</div>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">¶</a></h2>
<p>Please, see &lt;tax_DB_path&gt;/&lt;tax_DB_name&gt;/QIIME/problematic_taxa.txt file for identical sequences that had taxonomic disagreeing identifiers above the genus rank.</p>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">¶</a></h2>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id14" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Wang Q, Garrity GM, Tiedje JM, Cole JR. Naïve Bayesian classifier for rapid assignment of rRNA sequences into the new bacterial taxonomy. Appl Environ Microbiol. 2007.</p>
</aside>
<aside class="footnote brackets" id="id15" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>Caporaso JG, Kuczynski J, Stombaugh J, Bittinger K, Bushman FD, Costello EK, et al. QIIME allows analysis of high-throughput community sequencing data. Nature Methods. 2010.</p>
</aside>
<aside class="footnote brackets" id="id16" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">3</a><span class="fn-bracket">]</span></span>
<p>Murali A, Bhargava A, Wright ES. IDTAXA: A novel approach for accurate taxonomic classification of microbiome sequences. Microbiome. 2018.</p>
</aside>
<aside class="footnote brackets" id="id17" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id5">2</a>)</span>
<p>Bokulich, N. A., Kaehler, B. D., Rideout, J. R., Dillon, M., Bolyen, E., Knight, R., Huttley, G. A., &amp; Gregory Caporaso, J. (2018). Optimizing taxonomic classification of marker-gene amplicon sequences with QIIME 2’s q2-feature-classifier plugin. In Microbiome (Vol. 6, Issue 1). Springer Science and Business Media LLC. <a class="reference external" href="https://doi.org/10.1186/s40168-018-0470-z">https://doi.org/10.1186/s40168-018-0470-z</a></p>
</aside>
<aside class="footnote brackets" id="id18" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">5</a><span class="fn-bracket">]</span></span>
<p>Compeau PEC, Pevzner PA, Tesler G, Papoutsoglou G, Roscito JG, Dahl A, et al. Cutadapt removes adapter sequences from high-throughput sequencing reads kenkyuhi hojokin gan rinsho kenkyu jigyo. EMBnet.journal. 2013.</p>
</aside>
<aside class="footnote brackets" id="id19" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">6</a><span class="fn-bracket">]</span></span>
<p>Rognes T, Flouri T, Nichols B, Quince C, Mahé F. VSEARCH: A versatile open source tool for metagenomics. PeerJ. 2016</p>
</aside>
</aside>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">zAMP</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Installation and resource requirements</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Taxonomic reference databases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#database-processing-principle">Database processing principle</a></li>
<li class="toctree-l2"><a class="reference internal" href="#custom-database">Custom database</a></li>
<li class="toctree-l2"><a class="reference internal" href="#available-databases">Available databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Pipeline execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="under_the_hood.html">Under the hood</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="insilico_validation.html"><em>In silico</em> validation tool</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="setup.html" title="previous chapter">Installation and resource requirements</a></li>
      <li>Next: <a href="execution.html" title="next chapter">Pipeline execution</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2020, MetaGenLab.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/pages/ref_DB_preprocessing.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>