<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Under the hood &#8212; zAMP v.1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=6c7b0eaa"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Downstream Analysis" href="downstream_analysis.html" />
    <link rel="prev" title="Running zAMP" href="execution.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="under-the-hood">
<span id="id1"></span><h1>Under the hood<a class="headerlink" href="#under-the-hood" title="Link to this heading">¶</a></h1>
<section id="snakemake-environments-and-containers">
<h2>Snakemake, environments and containers<a class="headerlink" href="#snakemake-environments-and-containers" title="Link to this heading">¶</a></h2>
<p><a class="reference external" href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html">Snakemake</a> is the center-piece of this pipeline.
Snakemake is a Python-based workflow-manager that enables the processing of a large set of amplicon-based metagenomics sequencing reads into actionable outputs.
Each step is defined as a rule in which input/output files, software dependencies (Conda or containers), scripts and command-lines are specified (See <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html">snakemake’s docs</a> for more details).</p>
<p><a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html">Conda environments</a></p>
<p>Conda is a language-independent package and environment management tool. A Conda environment is a collection of installed Conda packages. For example, a research project might require VSEARCH 2.20.0 and its dependencies, whereas another environment associated with a completed project might necessitate the use of VSEARCH 2.15. Changing the environment, has no effect on the others.
Switching between environments is simple because they can be easily activated or deactivated.</p>
<p><a class="reference external" href="https://apptainer.org/docs/user/latest/">Apptainer containers</a></p>
<p>The concept of reproducible analysis in bioinformatics extends beyond good documentation and code sharing. Analyses typically depend on an entire environment with numerous tools, libraries, and settings.
Storage, reuse, and sharing environments via container software such as Docker and Singularity could improve reproducibility and productivity.
By using containers (apptainer, docker, podman …), users can create a single executable file that contains all aspects of their environment and allows to safely run environments from a variety of resources without requiring privileged access.</p>
</section>
<section id="logging-and-traceability">
<h2>Logging and traceability<a class="headerlink" href="#logging-and-traceability" title="Link to this heading">¶</a></h2>
<section id="logs">
<h3>logs<a class="headerlink" href="#logs" title="Link to this heading">¶</a></h3>
<p>Upon each execution, <em>zAMP</em> automatically creates a log file where all the standard output is recorded:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp_out</span><span class="o">/</span><span class="n">zamp</span><span class="o">.</span><span class="n">log</span>
</pre></div>
</div>
</section>
<section id="config-file">
<h3>config file<a class="headerlink" href="#config-file" title="Link to this heading">¶</a></h3>
<p>In addition to logs, <em>zAMP</em> copies a config file listing all the parameters used in the run unde</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zamp_out</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span>
</pre></div>
</div>
</section>
</section>
<section id="sequencing-reads-qc">
<h2>Sequencing reads QC<a class="headerlink" href="#sequencing-reads-qc" title="Link to this heading">¶</a></h2>
<p>QC rules assess the sequencing quality of all each sample with FastQC <a class="footnote-reference brackets" href="#id4" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>. Then, a MultiQC <a class="footnote-reference brackets" href="#id5" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a> report generates a report for each sequencing run (based on “run” column indicated  in <em>sample sheet</em> ).
A global MultiQC report is generated as well, but without interactive features to deal with the high number of samples</p>
</section>
<section id="denoising">
<h2>Denoising<a class="headerlink" href="#denoising" title="Link to this heading">¶</a></h2>
<section id="vsearch-otu-clustering">
<h3>Vsearch (OTU clustering)<a class="headerlink" href="#vsearch-otu-clustering" title="Link to this heading">¶</a></h3>
<section id="pandaseq">
<h4>PANDAseq<a class="headerlink" href="#pandaseq" title="Link to this heading">¶</a></h4>
</section>
<section id="vsearch">
<h4>Vsearch<a class="headerlink" href="#vsearch" title="Link to this heading">¶</a></h4>
</section>
</section>
<section id="dada2-asv-denoising">
<h3>DADA2 (ASV denoising)<a class="headerlink" href="#dada2-asv-denoising" title="Link to this heading">¶</a></h3>
<section id="cutadapt">
<h4>cutadapt<a class="headerlink" href="#cutadapt" title="Link to this heading">¶</a></h4>
</section>
<section id="dada2">
<h4>DADA2<a class="headerlink" href="#dada2" title="Link to this heading">¶</a></h4>
</section>
</section>
</section>
<section id="taxonomic-assignment">
<h2>Taxonomic assignment<a class="headerlink" href="#taxonomic-assignment" title="Link to this heading">¶</a></h2>
<section id="reference-database">
<h3>reference database<a class="headerlink" href="#reference-database" title="Link to this heading">¶</a></h3>
</section>
<section id="classifiers">
<h3>classifiers<a class="headerlink" href="#classifiers" title="Link to this heading">¶</a></h3>
</section>
</section>
<section id="post-processing">
<h2>Post-processing<a class="headerlink" href="#post-processing" title="Link to this heading">¶</a></h2>
<section id="taxonomic-filtering">
<h3>Taxonomic filtering<a class="headerlink" href="#taxonomic-filtering" title="Link to this heading">¶</a></h3>
</section>
<section id="rarefaction">
<h3>Rarefaction<a class="headerlink" href="#rarefaction" title="Link to this heading">¶</a></h3>
</section>
<section id="phylogenetic-tree-generation">
<h3>Phylogenetic tree generation<a class="headerlink" href="#phylogenetic-tree-generation" title="Link to this heading">¶</a></h3>
</section>
<section id="taxonomic-collapsing">
<h3>Taxonomic collapsing<a class="headerlink" href="#taxonomic-collapsing" title="Link to this heading">¶</a></h3>
</section>
<section id="normalization-and-abundance-based-filtering">
<h3>Normalization and abundance-based filtering<a class="headerlink" href="#normalization-and-abundance-based-filtering" title="Link to this heading">¶</a></h3>
</section>
<section id="exports">
<h3>Exports<a class="headerlink" href="#exports" title="Link to this heading">¶</a></h3>
</section>
<section id="fromatting">
<h3>Fromatting<a class="headerlink" href="#fromatting" title="Link to this heading">¶</a></h3>
<section id="wide-to-long-melting">
<h4>Wide to long melting<a class="headerlink" href="#wide-to-long-melting" title="Link to this heading">¶</a></h4>
</section>
<section id="transpose-and-meta-count-table">
<h4>transpose_and_meta_count_table<a class="headerlink" href="#transpose-and-meta-count-table" title="Link to this heading">¶</a></h4>
</section>
<section id="qiime2-formats">
<h4>Qiime2 formats<a class="headerlink" href="#qiime2-formats" title="Link to this heading">¶</a></h4>
</section>
</section>
</section>
<section id="picrust2">
<h2>Picrust2<a class="headerlink" href="#picrust2" title="Link to this heading">¶</a></h2>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">¶</a></h2>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">1</a><span class="fn-bracket">]</span></span>
<p>Andrews S, Krueger F, Seconds-Pichon A, Biggins F, Wingett S. FastQC. A quality control tool for high throughput sequence data. Babraham Bioinformatics. Babraham Institute. 2015.</p>
</aside>
<aside class="footnote brackets" id="id5" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">2</a><span class="fn-bracket">]</span></span>
<p>Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: Summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016;</p>
</aside>
</aside>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">zAMP</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Installation and resource requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="ref_DB_preprocessing.html">Taxonomic reference databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Running zAMP</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Under the hood</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#snakemake-environments-and-containers">Snakemake, environments and containers</a></li>
<li class="toctree-l2"><a class="reference internal" href="#logging-and-traceability">Logging and traceability</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequencing-reads-qc">Sequencing reads QC</a></li>
<li class="toctree-l2"><a class="reference internal" href="#denoising">Denoising</a></li>
<li class="toctree-l2"><a class="reference internal" href="#taxonomic-assignment">Taxonomic assignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#post-processing">Post-processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#picrust2">Picrust2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="downstream_analysis.html">Downstream Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="insilico_validation.html"><em>In silico</em> validation tool</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="execution.html" title="previous chapter">Running zAMP</a></li>
      <li>Next: <a href="downstream_analysis.html" title="next chapter">Downstream Analysis</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2020, MetaGenLab.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="../_sources/pages/under_the_hood.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>