## We recommend to format the taxonomy database to extract the region of gene of interest and merge the taxonomy of taxa that have
# identical sequences on this shorter gene region. However, in case where "extract_and_merge" is set to False in config file, these
# set of rules will be used to bypass the preprocessing of the database.
rule copy_fasta:
    input:
        os.path.join("{prefix}", "database", "original_seqs.fasta"),
    output:
        os.path.join("{prefix}", "database", "amplicons.fasta"),
    shell:
        """
        cp {input} {output}
        """


rule copy_taxonomy:
    input:
        os.path.join("{prefix}", "database", "clean_tax.tsv"),
    output:
        tax=os.path.join("{prefix}", "database", "amplicons_tax.tsv"),
        all_tax=os.path.join("{prefix}", "database", "amplicons_all_tax.tsv"),
    shell:
        """
        cp {input} {output.tax}
        awk 'BEGIN{{OFS=FS="\\t"}} NR==1{{$1="seq_id"; $2="all_tax"}} 1' {input} > {output.all_tax}
        """


rule Hash_unprocessed_fasta_tax:
    input:
        os.path.join("{prefix}", "database", "amplicons_tax.tsv"),
        os.path.join("{prefix}", "database", "amplicons_all_tax.txt"),
        os.path.join("{prefix}", "database", "amplicons.fasta"),
    output:
        os.path.join("{prefix}", "database", "fasta_tax.md5"),
    shell:
        """
        md5sum {input} | sort -k 2 | md5sum | cut -f 1 -d " " > {output}
        """
