<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Reproducible Scalable Pipeline For Amplicon-based Metagenomics (zAMP) &#8212; zAMP v.1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=6c7b0eaa"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation and resource requirements" href="pages/setup.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="reproducible-scalable-pipeline-for-amplicon-based-metagenomics-zamp">
<h1>Reproducible Scalable Pipeline For Amplicon-based Metagenomics (zAMP)<a class="headerlink" href="#reproducible-scalable-pipeline-for-amplicon-based-metagenomics-zamp" title="Link to this heading">¶</a></h1>
<p>zAMP is a bioinformatic pipeline designed for convenient, reproducible and scalable amplicon-based metagenomics. zAMP is a compilation of command-line tools and R packages that processes next-generation sequencing (NGS) from amplicon-based metagenomics reads into taxonomic profiles. Scripts running these tools are embedded in a <em>Snakemake</em> pipeline for reproducible, scalable and easy “<em>one-command-line</em>” execution of all steps, from QC to basic plotting. Tools and packages are made available into <em>Singularity</em> containers and <em>Conda</em> environments for easy and reproducible management of dependencies. zAMP is capable of processing local sequencing reads or reads stored on the Sequence Reads Archive (SRA). Complementary workflows enables preprocessing of reference taxonomic databases as well as <em>in silico</em> prediction the taxonomic classification obtained from reference genomes.</p>
<p>Please note that zAMP is a compilation of a number of published tools. We would be very grateful that you acknowledge our work but also the contribution of the developers of these tools in your publications using zAMP.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/setup.html">Installation and resource requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/setup.html#quick-start">Quick start</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/setup.html#details">Details</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/ref_DB_preprocessing.html">Taxonomic reference databases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#database-processing-principle">Database processing principle</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#custom-database">Custom database</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#available-databases">Available databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#output">Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/ref_DB_preprocessing.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/execution.html">Pipeline execution</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#select-a-method-to-define-sequencing-read-files-as-input">1. Select a method to define sequencing read files as input</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#create-a-working-directory">2. Create a working directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#create-a-links-directory-for-sample-and-oldsamplename-column-as-input">3. Create a <em>links</em> directory (<em>for Sample and OldSampleName column as input</em>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#write-a-sample-sheet-aka-metadata-table">4. Write a <em>sample sheet</em> (aka <em>metadata table</em>):</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#write-a-config-file">5. Write a <em>config file</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#make-sure-snakemake-is-available">6. Make sure Snakemake is available</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/execution.html#run-the-pipeline">7. Run the pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/under_the_hood.html">Under the hood</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#snakemake-environments-and-containers">Snakemake, environments and containers</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#input-output-definition"><strong>Input output definition</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#logging-and-traceability">Logging and traceability</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#sequencing-reads-qc">Sequencing reads QC</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#denoising">Denoising</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#taxonomic-assignment">Taxonomic assignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#post-processing">Post-processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#picrust2">Picrust2</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/under_the_hood.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/insilico_validation.html"><em>In silico</em> validation tool</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/insilico_validation.html#aim">Aim</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/insilico_validation.html#working-principle">Working principle</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/insilico_validation.html#inputs">Inputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/insilico_validation.html#execution">Execution</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/insilico_validation.html#output">Output</a></li>
</ul>
</li>
</ul>
</div>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">zAMP</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/setup.html">Installation and resource requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/ref_DB_preprocessing.html">Taxonomic reference databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/execution.html">Pipeline execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/under_the_hood.html">Under the hood</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/FAQ.html">Frequently asked questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/insilico_validation.html"><em>In silico</em> validation tool</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="pages/setup.html" title="next chapter">Installation and resource requirements</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2020, MetaGenLab.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>